xds/src/main/java/io/grpc/xds/XdsNameResolver.java

```diff
@@ -444,6 +560,198 @@ public void run() {
         });
       }
     }
+
+    @Nullable
+    private Long determineFaultDelayNanos(FaultDelay faultDelay, Metadata headers) {
+      Long delayNanos;
+      FractionalPercent fractionalPercent = faultDelay.percent();
+      if (faultDelay.headerDelay()) {
+        try {
+          int delayMillis = Integer.parseInt(headers.get(HEADER_DELAY_KEY));
+          delayNanos = TimeUnit.MILLISECONDS.toNanos(delayMillis);
+          String delayPercentageStr = headers.get(HEADER_DELAY_PERCENTAGE_KEY);
+          if (delayPercentageStr != null) {
+            int delayPercentage = Integer.parseInt(delayPercentageStr);
+            if (delayPercentage >= 0 && delayPercentage < fractionalPercent.numerator()) {
+              fractionalPercent =
+                  FractionalPercent.create(delayPercentage, fractionalPercent.denominatorType());
+            }
+          }
+        } catch (NumberFormatException e) {
+          return null; // treated as header_delay not applicable
+        }
+      } else {
+        delayNanos = faultDelay.delayNanos();
+      }
+      if (random.nextInt(1_000_000) >= getRatePerMillion(fractionalPercent)) {
+        return null;
+      }
+      return delayNanos;
+    }
+
+    @Nullable
+    private Status determineFaultAbortStatus(FaultAbort faultAbort, Metadata headers) {
+      Status abortStatus = null;
+      FractionalPercent fractionalPercent = faultAbort.percent();
+      if (faultAbort.headerAbort()) {
+        try {
+          String grpcCodeStr = headers.get(HEADER_ABORT_GRPC_STATUS_KEY);
+          if (grpcCodeStr != null) {
+            int grpcCode = Integer.parseInt(grpcCodeStr);
+            abortStatus = Status.fromCodeValue(grpcCode);
+          }
+          String httpCodeStr = headers.get(HEADER_ABORT_HTTP_STATUS_KEY);
+          if (httpCodeStr != null) {
+            int httpCode = Integer.parseInt(httpCodeStr);
+            abortStatus = GrpcUtil.httpStatusToGrpcStatus(httpCode);
+          }
+          String abortPercentageStr = headers.get(HEADER_ABORT_PERCENTAGE_KEY);
+          if (abortPercentageStr != null) {
+            int abortPercentage =
+                Integer.parseInt(headers.get(HEADER_ABORT_PERCENTAGE_KEY));
+            if (abortPercentage >= 0 && abortPercentage < fractionalPercent.numerator()) {
```

Comment:
Envoy doc says the value is _capped by_ configuration's numerator. So you shouldn't throw away the percentage even if the header value is greater than configuration's value. That's just 100%.

Same for delay's percentage.

Btw, the doc didn't mention if it should fallback to use configuration's abort percentage if the header doesn't exist. But I think that's a reasonable assumption.

Reply:
> Envoy doc says the value is capped by configuration's numerator. So you shouldn't throw away the percentage even if the header value is greater than configuration's value. That's just 100%.

If the header value is greater than the config's value, then the effective value is the config value. That's what envoy's [implementation](https://github.com/envoyproxy/envoy/blob/v1.17.0/source/extensions/filters/common/fault/fault_config.cc#L33) does.

Reply:
You are right.